Add .codegen.rewriting module

bjodah opened the pull request:
This module is useful for rewriting symbolic expressions to target specific floating point functions in e.g. C99.

bjodah committed:
Add .codgen.rewriting module

bjodah committed:
Use SymPy's XFAIL, minor enhancements to tests

smichr commented on sympy/codegen/rewriting.py:
```diff
@@ -0,0 +1,197 @@
+expressions for this purpose.
+
+Examples
+--------
```
Should this be `========` with a blank line following?

smichr commented on sympy/codegen/rewriting.py:
```diff
@@ -0,0 +1,197 @@
+    as argument.
+
+    Parameters
+    ----------
```
`==========`?

smichr replied:
ditto below

bjodah committed:
Address review (doc & compatibility)

- Added filterfalse to sympy.core.compatiblity
- Fixed formatting of docstrings in .codegen.rewriting
- Ensured doctests are run on module level docstring of .codegen.rewriting

bjodah commented:
Thanks @smichr 
I've always been a bit confused regarding `====` vs. `-----` (numpydoc [guidelines](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) are in favor of the latter but you're right that SymPy usually uses the former).

I haven added the module to `doc/src/modules/codegen.rst` since that would just give a merge conflict with #13100, but once one of these are merged I'll merge master in and add it.

asmeurer commented:
This looks like a good start.
